src/server/frontend_xwayland/xwayland_server.cpp

```diff
@@ -64,7 +64,12 @@ mf::XWaylandServer::~XWaylandServer()
       if (xserver_status == RUNNING)
         wm->destroy();
 
-      kill(pid, SIGTERM);
+      if (kill(pid, SIGTERM) == 0)
+      {
+          usleep(100000);           // After 100ms...
+          if (kill(pid, 0) == 0)    // ...if Xwayland is still running...
+            kill(pid, SIGKILL);     // ...then kill it!
```

Comment:
Is there a known reason SIGTERM isn't working?

Reply:
Empirically, Xwayland doesn't exit while there's an active X11 client.

Reply:
I question if a 0.1s delay to shutdown in all cases is an ideal solution. If SIGTERM is expected to often fail, and SIGKILL is harmless, why not just do that in the first place?

Reply:
Of course, there's no guarantee that `pid` still refers to the xwayland process… :grin: 

Unfortunately, this is wandering into the parts of the system where the API is so bad it is *impossible* to do the right thing. The *right* thing to do is to monitor the process we spawn and wait for it to die, but that requires signal handling that we can't rely on working because we're a library.

SIGTERM→SIGKILL seems like a reasonable kludge for now.

Reply:
`systemd-run` is pretty close to useful, but not quite.

Reply:
It is indeed possible that the pid has been reused, and even possible we're allowed to send signals to the new process.

But while racy it isn't as bad as it might be: The thread we're about to join _does_ use signal handling (via `waitpid()`) to monitor the Xwayland process and sets `xserver_status` (which we've just read) accordingly.

The code in this component needs further tidy up, which is why we default to disabling it behind an "experimental" configuration option. The plan is to incrementally improve it to the point where things reliably work as expected. Hopefully, we will be able to identify X11 clients properly to ask them politely to shut down before SIGKILLing anything.